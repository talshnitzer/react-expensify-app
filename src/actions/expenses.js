import uuid from 'uuid';
import database from '../firebase/firebase';

//before using firebase db the flow was:
//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//inorder to use firebase the flow changes to:
//components calls action generator
//action generator returns function
//component dispatches function (using rudux tool - redux-thunk)
//function runs (has ability to dispatch other actions an do whatever it wants)


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//old version of ADD_EXPENSE (before adding firebase calls)
// export const addExpense = (
//     { 
//         description  = '', 
//         note = '', 
//         amount = 0, 
//         createdAt = 0
//     } = {}
//     ) => ({
//     type: 'ADD_EXPENSE',
//     expense: {
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });


//this function gets called internally by redux, it gets called with dispatch.
//we write some data to firebase, waiting for that data to correctly sync,
//then we'll use dispatch to dispatch addExpense so the redux store reflects those changes as well.
export const startAddExpense = (expenseData = {}) => { //I get some expenseData as an argument, and if I don't I set it equal to an empty object.
    return (dispatch, getState) => {             //this will work because we installed the 'thunk' middleware. w/o it it will not work. this gives us access to dispatch and get state fundtions
                                        //returning a function instead of an object
        const uid = getState().auth.uid;
        const {
        description  = '', 
        note = '', 
        amount = 0, 
        createdAt = 0
        } = expenseData; //distructure from expenseData
        const expense = {description, note, amount, createdAt};

        //by returning the promise chain, we can continue chaining on over at expenses.test.js
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({id}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({id}));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({   //arguments ID and updates, neither of which need defaults if you don't
                                        //have the ID and you're not updating anything there's no need to call it.
                                        //And then we'll implicitly return an object.
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return(dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses`) //The 'return' here, makes sure the promise actually gets returned, and thats allow us to have access to 'then' at app.js were we dispatch.
            .once('value')
            .then((snapshot) => {
                const expenses = []; 

                snapshot.forEach((childSnapshot) => {
                 expenses.push({
                     id: childSnapshot.key,
                     ...childSnapshot.val()
                 });
                });
                dispatch(setExpenses(expenses));
             });
    };
}; //'startSetExpenses' is the async action - fetch the data from db and than dispatch 'setExpenses'
