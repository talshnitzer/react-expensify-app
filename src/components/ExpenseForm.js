import React from "react";
//'Moment' is a time utility library. it makes it asy to work with time.
import moment from 'moment';
//'react-dates' is a calendar picker tool that require moment.
import { SingleDatePicker } from 'react-dates';


//using 'class' based component because we use component 'state'.
export default class ExpenseForm extends React.Component {
    //'placeholder'- something to show in the input when there is no text typed by the user
    //'autoFocus'- when we visit the page it automatically puts the cursor and the focus on that input
    //use local component state to track the changes to all of these inputs,
    //submit form will actually send that redx to edit the existing expense or create a new one

    //only use the defaults for these four state values.
    //If no expense was passed down. so the AddExpense page still works,
    //But if an expense was passed down we want to start these off at those values.
    //we have to use constructor function for that

    constructor(props) {
        super(props);

        this.state={
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/ 100).toString() : '',
            //if 'expense' exist, create an instance of moment at a specific point in time,
            //so we pass to the moment() func a timestatmp
            createdAt: props.expense ? moment(props.expense.createdAt): moment(),
            //focused means that the calendar is open
            calendarFocused: false,
            error: ''
        };
    }
    
    //the updater function in 'setState':
    //implicitly return an object, So I wrap it in parentheses,
    //set 'description = description variable', so I can use the ES6 object shorthand.
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState (() => ({note}));
    };

    onAmountChange = (e) => {
        const amount =e.target.value;
        //regular expression to enforce the 'amount' format. details on the botom.
        //'!amount' condition added in order to allow the user to delete the amount entered the the input
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }  
    };

    //in  the API- 'onDateChange' gets called with the moment. we call it 'createdAt'
    onDateChange = (createdAt) => {
        if (createdAt) { //to prevent the user from deleting the date fro the input
            this.setState(() => ({ createdAt }));
        } 
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({calendarFocused: focused}));
    
    }

    onSubmit = (e) => {
        e.preventDefault(); //so there will not be a full page refresh, but handle this with js instead
        if (!this.state.description || !this.state.amount) {  
            this.setState(() => ({error: 'Please provide description and amount'}));
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({ //here we pass the 'expense' object to EditExpense and AddExpense components where the 'dispatch' is done
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, //base 10, convert to cents
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
                <form className="form" onSubmit={this.onSubmit}>
                    {this.state.error && <p className="form__error">{this.state.error}</p>} 
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        className="text-input"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        //customize number of months see,and that days in the past will be available.
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea 
                    placeholder="Add a note for your expense (optional)"
                    className="textarea"
                    value={this.state.text}
                    onChange={this.onNoteChange}
                    >
                    </textarea>
                    <div> 
                        <button className="big_button">Save Expense</button>
                    </div>
                </form>
        )
    }
}

//why 'button' element above has its own div? this is for styling purposes:
//the 'form' element has a display 'flex' that cause the direct elements under it to take 
//the full width of the screen. we don't want that for the 'button' element

//REGEX build explanation:
//look for a digit and we use backslash 'd' for that.
//we don't want to match a letter upfront So we force the string to start with a digit by the '^' key.
//to allow the user to add many digits - we use '{1,}' saying we want range from 1 to infinity digits.
//to add a decimal point, optional, you could have a decimal maybe you don't.
//But if I do have it I want to make sure it's actually a match.
// to create an optional group: set up parentheses and a question mark at the end.
//what we've done is we've created a new group matching it between 0 and 1 times.
//That means you don't have to have this at all or if you do have it you just have it once which is exactly
//inside the parenthesis- the escape that with a backslash and DOT character.
//to match it literally then we're looking for numbers once again.
//We want to allow for 0 to 2 digits after the dot.
//we use curly braces to define that range on the low end zero on the high end 2.
//at the end a dollar sign.
//This makes sure that the regular expression ends with what we have here as opposed to having other characters

