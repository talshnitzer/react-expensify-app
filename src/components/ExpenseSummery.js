import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/selectExpensesTotal';
import numeral from 'numeral';

//Tal's solution
// export const ExpenseSummery = (props) => (
//     <div>
//         <p>Viewing {props.expenses.length} expenses totalling {numeral(selectExpensesTotal(props.expenses) / 100).format('$0,0.00')}</p>
//     </div>
// );

//Andrew's solution
export const ExpenseSummery = ({expenseCount, ExpensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpenseTotal = numeral(ExpensesTotal / 100).format('$0,0.00');

    return (
        <div className="page_header">
            <div className="content-container">
                <h1 className="page_header__title">Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpenseTotal}</span></h1>
                <div className="page-header__actions">
                    <Link className="big_button" to="/create">Add Expense</Link>
                </div>
            </div> 
        </div>
    );
};

//Tal's sollution
// const mapStateToProps = (state) => {
//     return{
//         expenses: selectExpenses(state.expenses, state.filters)
//     };
// };

//Andrew's solution
const mapStateToProps = (state) => {
    const visibleExpenses =selectExpenses(state.expenses, state.filters);
    return{
        expenseCount: visibleExpenses.length,
        ExpensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect (mapStateToProps)(ExpenseSummery);