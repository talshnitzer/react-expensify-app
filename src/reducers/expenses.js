//Expense Reducer

const expensesReducerDefaultState = [];

//use a default export since there is  one thing to export is the reducer function
export default  (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //return state.filter(expense => expense.id !== action.expense.id);
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return{                        //return a new object changing something:
                        ...expense,                 //grab all of its existing properties.
                        ...action.updates           //override any of the ones that were passed down in 'action.updates' and that's the new value for that expense                  
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};

