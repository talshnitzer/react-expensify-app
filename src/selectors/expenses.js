import moment from 'moment';

//Get visible expenses
export default  (expenses, {text, sortBy, startDate, endDate}) => { //destructurin the filters object
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') :  true ;  //is it the same day or before?
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment,'day') : true;
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a,b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1; //1=> b comes first, -1=>a comes first
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1 ;
        }
    });
};