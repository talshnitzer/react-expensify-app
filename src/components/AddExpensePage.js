import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

//confeting the stateless function componenet into a class, 
//so We no longer define the function 'onSubmit' inline,
// which would need to get recalculated on every render.
//the former staeless function + all comments explaining the code, is commented out below.

export class AddExpensePage extends React.Component {
    onSubmit = (expense)=> {
        this.props.addExpense(expense); 
        this.props.history.push('/'); 
    };
    render() {
        return (
            <div>
    <h1>Add Expense</h1>
    <ExpenseForm 
        onSubmit={this.onSubmit}
    />
    </div>
        )
    };
};

 //connecting the dispatch call to addExpense, from outside to the AddExensePage for testing purpose
 const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
 });

 export default connect(undefined, mapDispatchToProps)(AddExpensePage);

//  const AddExpensePage = (props) => (
//     <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm 
//         onSubmit={(expense)=> {
//             //props.dispatch(addExpense(expense));//adding the data to redux store
//             props.onSubmit(expense); //the above line change done for testing porpose
//            // components rendered inside of react router  get access to special props like 'history' below.
//             props.history.push('/'); //switch over to dashboard page as if I clicked the link, not through the full page refresh
//         }}
//     />
//     </div>
//  );