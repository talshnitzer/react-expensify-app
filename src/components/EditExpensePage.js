import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense,  startRemoveExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => { //the expense we get from the form
        //this.props.editExpense(expense); //Tal version of solution to testing
        this.props.startEditExpense(this.props.expense.id, expense); //Andrew's version of solution to testing
        this.props.history.push('/'); //get back to dashboard
    };
    onClick = () => {
        //this.props.removeExpense(); //Tal version of solution to testing
        this.props.startRemoveExpense({id: this.props.expense.id}); //Andrew's version of solution to testing
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className="page_header">
                    <div className="content-container">
                    <h1 className="page_header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm   
                    expense={this.props.expense} //the existing expense
                    onSubmit={this.onSubmit}
                    ></ExpenseForm>
                    <button className="big_button button--secondary" onClick = {this.onClick}>Remove Expense</button>
                </div>
            </div>
        )
    };
};

const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});


//Andrew's version of solution to testing
const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});


//Tal's version of solution to testing
// const mapDispatchToProps = (dispatch,props) => {
//     return {
//         editExpense: (expense) => dispatch(editExpense(props.match.params.id,{...expense})),
//         removeExpense: () => dispatch(removeExpense({id: props.match.params.id}))
//     };
// };

 export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

 //We're searching the expenses array for an expense whose I.D. matches Prop.match.params.ID.
// how do we access it?
//we have access to the prop's at The second argumen of mapStateToProps
//we can take some of the current props that were passed into the high order component and we can use
//them to calculate the props that we want to add on.
// so react-router renders our higher order component,
//the higher order component passes the props through,
// and it also allows us to add on some new ones

