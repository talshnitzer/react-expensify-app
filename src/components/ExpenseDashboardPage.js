import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummery from './ExpenseSummery';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummery />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
 );

export default ExpenseDashboardPage;