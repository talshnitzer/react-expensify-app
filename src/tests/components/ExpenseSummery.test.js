import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummery } from '../../components/ExpenseSummery';
import expenses from '../fixtures/expenses';

test('should render ExpenseSummery correctly with 1 expense' , () => {
    const wrapper = shallow(<ExpenseSummery expenses={[expenses[0]]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummery correctly with 2 expenses' , () => {
    const wrapper = shallow(<ExpenseSummery expenses={[expenses[0], expenses[1]]}/>);
    expect(wrapper).toMatchSnapshot();
});

