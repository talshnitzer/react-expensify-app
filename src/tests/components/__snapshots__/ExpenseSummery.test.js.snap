// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummery correctly with 1 expense 1`] = `
<div
  className="page_header"
>
  <div
    className="content-container"
  >
    <h1
      className="page_header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="big_button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummery correctly with multiple expenses 1`] = `
<div
  className="page_header"
>
  <div
    className="content-container"
  >
    <h1
      className="page_header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling 
      <span>
        $235,123,468.90
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="big_button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
